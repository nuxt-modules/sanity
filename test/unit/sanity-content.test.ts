import { expect, describe, it } from 'vitest'
import { mount } from '@vue/test-utils'
import SanityContent from '../../src/runtime/components/sanity-content'

type ExampleBlock = { _type: string; [key: string]: any }

const exampleBlocks: Record<string, ExampleBlock | ExampleBlock[]> = {
  link: {
    _type: 'link',
    href: 'http://test.com/',
  },
  heading: {
    _key: 'b1905c55df85',
    _type: 'block',
    children: [
      {
        _key: 'b1905c55df850',
        _type: 'span',
        marks: [],
        text: '3. Providing your personal data to others',
      },
    ],
    markDefs: [],
    style: 'h3',
  },
  list: [
    {
      _type: 'block',
      _key: 'a6cec1a2a738',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'b39d8ec32f27',
          text: 'test',
          marks: [],
        },
      ],
      level: 1,
      listItem: 'number',
    },
    {
      _type: 'block',
      _key: '5917a3f6485d',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'aaa0e0ab720a',
          text: 'thing',
          marks: [],
        },
      ],
      level: 1,
      listItem: 'number',
    },
  ],
  nestedList: [
    {
      _key: 'b1905c55df85',
      _type: 'block',
      children: [
        {
          _key: 'b1905c55df850',
          _type: 'span',
          marks: [],
          text: '3. Providing your personal data to others',
        },
      ],
      markDefs: [],
      style: 'h3',
    },
    {
      _type: 'block',
      _key: 'a6cec1a2a738',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'b39d8ec32f27',
          text: 'test',
          marks: [],
        },
      ],
      level: 1,
      listItem: 'number',
    },
    {
      _type: 'block',
      _key: '5917a3f6485d',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'aaa0e0ab720a',
          text: 'thing',
          marks: [],
        },
      ],
      level: 1,
      listItem: 'number',
    },
    {
      _type: 'block',
      _key: 'ceaddc3e7d34',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: '62d7844aaf8e',
          text: 'nested',
          marks: [],
        },
      ],
      level: 2,
      listItem: 'bullet',
    },
    {
      _type: 'block',
      _key: '415d8fb2fa1e',
      style: 'normal',
      markDefs: [],
      children: [
        {
          _type: 'span',
          _key: 'a7705718303a',
          text: 'list',
          marks: [],
        },
      ],
      level: 2,
      listItem: 'bullet',
    },
  ],
  blockquote: {
    _key: 'd810da8ac845',
    _type: 'block',
    children: [
      {
        _key: 'd810da8ac8450',
        _type: 'span',
        marks: [],
        text: 'Build your next Vue.js application with confidence using ',
      },
      {
        _key: 'd810da8ac8451',
        _type: 'span',
        marks: [
          'c9c6224401ba',
        ],
        text: 'NuxtJS',
      },
      {
        _key: 'd810da8ac8452',
        _type: 'span',
        marks: [],
        text: '. An open source framework making web development simple and powerful.',
      },
    ],
    markDefs: [
      {
        _key: 'c9c6224401ba',
        _type: 'link',
        href: 'https://nuxtjs.org',
      },
    ],
    style: 'blockquote',
  },
  customBlockTypes: {
    _key: 'd810da8ac845',
    _type: 'block',
    children: [
      {
        _key: 'd810da8ac8450',
        _type: 'span',
        marks: [],
        text: 'An example of ',
      },
      {
        _key: '7dc51c030a2f',
        _type: 'customIcon',
        icon: 'cog',
        text: 'custom',
      },
      {
        _key: '230498a',
        _type: 'span',
        marks: [],
        text: ' block types.',
      },
    ],
    markDefs: [],
    style: 'normal',
  },
  strong: {
    _key: 'd810da8ac845',
    _type: 'block',
    children: [
      {
        _key: 'd810da8ac8450',
        _type: 'span',
        marks: [],
        text: 'An example of ',
      },
      {
        _key: '7dc51c030a2f',
        _type: 'span',
        marks: [
          'strong',
        ],
        text: 'bold',
      },
      {
        _key: '794714489c2d',
        _type: 'span',
        marks: [],
        text: ' text and ',
      },
      {
        _key: '23487f',
        _type: 'span',
        marks: [
          'strong',
        ],
        text: 'another',
      },
      {
        _key: '230498a',
        _type: 'span',
        marks: [],
        text: '.',
      },
    ],
    markDefs: [],
    style: 'normal',
  },
  exampleMarkDefs: {
    _key: '3522a2a863b9',
    _type: 'block',
    children: [
      {
        _key: '3522a2a863b90',
        _type: 'span',
        marks: ['strong'],
        text: 'Google Analytics',
      },
      {
        _key: '3522a2a863b91',
        _type: 'span',
        marks: [],
        text:
          ': The Sites use Google Analytics. Further information is available at ',
      },
      {
        _key: '3522a2a863b92',
        _type: 'span',
        marks: ['71968b05862f'],
        text: 'https://tools.google.com/dlpage/gaoptout/',
      },
      {
        _key: '3522a2a863b93',
        _type: 'span',
        marks: [],
        text: '.',
      },
    ],
    level: 2,
    listItem: 'bullet',
    markDefs: [
      {
        _key: '71968b05862f',
        _type: 'link',
        href: 'https://tools.google.com/dlpage/gaoptout/',
      },
    ],
    style: 'normal',
  },
}

describe('SanityContent', () => {
  it('should render with no props', () => {
    const wrapper = mount(SanityContent)
    expect(wrapper.html()).toMatchSnapshot()
  })

  Object.entries(exampleBlocks).forEach(([component, block]) => {
    it(`should render ${component} blocks`, () => {
      const wrapper = mount(SanityContent, {
        props: {
          blocks: Array.isArray(block) ? block : [block],
          serializers: {
            types: { customIcon: 'i' },
          },
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
