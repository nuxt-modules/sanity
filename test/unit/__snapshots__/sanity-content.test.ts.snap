// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SanityContent > should render adjacentLists blocks 1`] = `
"<ul>
  <li>item</li>
  <li>item</li>
  <li>item</li>
</ul>
<ol>
  <li>item</li>
  <li>item</li>
  <li>item</li>
</ol>"
`;

exports[`SanityContent > should render blockquote blocks 1`] = `"<blockquote>Build your next Vue.js application with confidence using <a href=\\"https://nuxtjs.org\\">NuxtJS</a>. An open source framework making web development simple and powerful.</blockquote>"`;

exports[`SanityContent > should render customBlockTypes blocks 1`] = `"<p>An example of <i>custom</i> block types.</p>"`;

exports[`SanityContent > should render customComponent blocks 1`] = `
"<div>propValuetext content</div>
<div>propValuetext content</div>"
`;

exports[`SanityContent > should render evenMoreNestedList blocks 1`] = `
"<ol>
  <li>1. Top level</li>
  <li>
    <ol>
      <li>a. Second level</li>
      <li>
        <ol>
          <li>i. Third level</li>
        </ol>
      </li>
      <li>b. Second level</li>
    </ol>
  </li>
  <li>2. Top level</li>
  <li>
    <ol>
      <li>a. Second level</li>
    </ol>
  </li>
</ol>"
`;

exports[`SanityContent > should render exampleMarkDefs blocks 1`] = `
"<ul>
  <li><strong>Google Analytics</strong>: The Sites use Google Analytics. Further information is available at <a href=\\"https://tools.google.com/dlpage/gaoptout/\\">https://tools.google.com/dlpage/gaoptout/</a>.</li>
</ul>"
`;

exports[`SanityContent > should render heading blocks 1`] = `"<h3>3. Providing your personal data to others</h3>"`;

exports[`SanityContent > should render link blocks 1`] = `"<a href=\\"http://test.com/\\"></a>"`;

exports[`SanityContent > should render list blocks 1`] = `
"<ol>
  <li>test</li>
  <li>thing</li>
</ol>"
`;

exports[`SanityContent > should render multipleDecorators blocks 1`] = `"<p>with <strong><em>bold and italic</em></strong> decorators</p>"`;

exports[`SanityContent > should render nestedList blocks 1`] = `
"<h3>3. Providing your personal data to others</h3>
<ol>
  <li>test</li>
  <li>thing</li>
  <li>
    <ul>
      <li>nested</li>
      <li>list</li>
    </ul>
  </li>
</ol>"
`;

exports[`SanityContent > should render strong blocks 1`] = `"<p>An example of <strong>bold</strong> text and <strong>another</strong>.</p>"`;

exports[`SanityContent > should render textWithProps blocks 1`] = `
"<div>hello worldIâ€™m a heading!</div>
<blockquote class=\\"blockquote\\">Build your next Vue.js application with confidence using <a href=\\"https://nuxtjs.org\\">NuxtJS</a>. An open source framework making web development simple and powerful.</blockquote>
<h1>What we do</h1>"
`;

exports[`SanityContent > should render with no props 1`] = `""`;
